common/djangoapps/third_party_auth/models.py

```diff
@@ -0,0 +1,385 @@
+    """
+    General configuration required for this edX instance to act as a SAML
+    Service Provider and allow users to authenticate via third party SAML
+    Identity Providers (IdPs)
+    """
+    private_key = models.TextField(
+        help_text=(
+            'To generate a key pair as two files, run '
+            '"openssl req -new -x509 -days 3652 -nodes -out saml.crt -keyout saml.key". '
+            'Paste the contents of saml.key here.'
+        )
+    )
+    public_key = models.TextField(help_text="Public key certificate.")
+    entity_id = models.CharField(max_length=255, default="http://saml.example.com", verbose_name="Entity ID")
+    org_info_str = models.TextField(
+        verbose_name="Organization Info",
+        default='{"en-US": {"url": "http://www.example.com", "displayname": "Example Inc.", "name": "example"}}',
+        help_text="JSON dictionary of 'url', 'displayname', and 'name' for each language",
+    )
+    other_config_str = models.TextField(
+        default='{\n"SECURITY_CONFIG": {"metadataCacheDuration": 604800, "signMetadata": false}\n}',
+        help_text=(
+            "JSON object defining advanced settings that are passed on to python-saml. "
+            "Valid keys that can be set here include: SECURITY_CONFIG, SP_NAMEID_FORMATS, SP_EXTRA"
+        ),
+    )
+
+    class Meta(object):  # pylint: disable=missing-docstring
+        verbose_name = "SAML Configuration"
+        verbose_name_plural = verbose_name
+
+    def clean(self):
+        """ Standardize and validate fields """
+        super(SAMLConfiguration, self).clean()
+        self.org_info_str = clean_json(self.org_info_str, dict)
+        self.other_config_str = clean_json(self.other_config_str, dict)
+
+        self.private_key = self.private_key.replace("-----BEGIN PRIVATE KEY-----", "").strip()
+        self.private_key = self.private_key.replace("-----END PRIVATE KEY-----", "").strip()
+        self.public_key = self.public_key.replace("-----BEGIN CERTIFICATE-----", "").strip()
+        self.public_key = self.public_key.replace("-----END CERTIFICATE-----", "").strip()
+
+    def get_setting(self, name):
+        """ Get the value of a setting, or raise KeyError """
+        if name == "ORG_INFO":
+            return json.loads(self.org_info_str)
+        if name == "SP_ENTITY_ID":
+            return self.entity_id
+        if name == "SP_PUBLIC_CERT":
+            return self.public_key
+        if name == "SP_PRIVATE_KEY":
+            return self.private_key
+        if name == "TECHNICAL_CONTACT":
+            return {"givenName": "Technical Support", "emailAddress": settings.TECH_SUPPORT_EMAIL}
+        if name == "SUPPORT_CONTACT":
+            return {"givenName": "SAML Support", "emailAddress": settings.TECH_SUPPORT_EMAIL}
+        other_config = json.loads(self.other_config_str)
+        return other_config[name]  # SECURITY_CONFIG, SP_NAMEID_FORMATS, SP_EXTRA
+
+
+class SAMLProviderData(models.Model):
```

Comment:
I did actually code it that way at first (using `ConfigurationModel`). But of the three columns that ConfigurationModel defines, I didn't really want any of them for this model. `change_date` is ambiguous and has auto_now_add=True, whereas in the model I wrote here we have a `fetch_date` which is more clear and which gets updated to the current time during every fetch where there are no changes to the data. `changed_by` doesn't apply, and `enabled` isn't necessary since providers should be enabled/disabled via the `SAMLProviderConfig` model instead.

So it seemed like ConfigurationModel would add things we didn't need, and didn't provide that much in the way of functionality code for this. It is also arguably confusing to use ConfigurationModel for something that isn't "configured" by the user.